lib/execute.js

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2017 Resin.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const debug = require('debug')(require('../package.json').name);
+const Bluebird = require('bluebird');
+const fs = Bluebird.promisifyAll(require('fs'));
+const childProcess = require('child_process');
+const path = require('path');
+const tmp = require('tmp');
+tmp.setGracefulCleanup();
+
+/**
+ * @summary Run a platform script
+ * @function
+ * @public
+ *
+ * @param {String} script - path to script
+ * @param {Function} callback - callback (error, output)
+ *
+ * @example
+ * scripts.run(scripts.paths.win32, (error, output) => {
+ *   if (error) {
+ *     throw error;
+ *   }
+ *
+ *   console.log(output);
+ * });
+ */
+exports.extractAndRun = (script, callback) => {
+  tmp.file({
+    postfix: path.extname(script.filename),
+
+    // 0755 to decimal, because strict
+    // mode disallows octal literals
+    mode: 493
+
+  }, (error, temporaryPath, fileDescriptor) => {
+    if (error) {
+      return callback(error);
+    }
+
+    fs.writeFileAsync(fileDescriptor, script.content).return(fileDescriptor).then(fs.closeAsync).then(() => {
+      return new Bluebird((resolve, reject) => {
+
+        // On Windows, if the directory containing the temporary path
+        // includes an ampersand (e.g. if the username contains an ampersand)
+        // then the execution will fail.
+        // Turns out we can pass the temporary script dirname as the current
+        // working directory and execute the basename, and then it all works.
+        return childProcess.execFile(path.basename(temporaryPath), {
```

Comment:
Now we're using this new self-extracting script mechanism, can't we just change this to `.exec` instead of `.execFile`, which _does_ handle ampersands correctly?